File: tests/python/test_febio.py
"""FEBio input files (``.feb``): both engines, the fixtures written by
``tools/gen_febio_fixtures.py`` in FEBio's own node numbering (spec 2.5, 3.0
and 4.0), and the spec-4.0 writer."""

import pathlib

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus import _core
from meshioplusplus.febio import _febio as py_febio

from .test_node_order import _is_valid

MESHES = pathlib.Path(__file__).parent / "meshes" / "febio"
FIXTURES = sorted(MESHES.glob("*.feb"))


@pytest.fixture(params=["core", "python"])
def engine(request):
    """Both engines behind the same read/write signatures."""
    if request.param == "core":
        return meshioplusplus.febio
    return py_febio


def _regions(mesh):
    return sorted(
        (r.kind, r.name, r.dim, r.tag, tuple(np.asarray(r.entries).ravel()))
        for r in mesh.regions
    )


def _same(a, b):
    np.testing.assert_array_equal(a.points, b.points)
    assert [c.type for c in a.cells] == [c.type for c in b.cells]
    for x, y in zip(a.cells, b.cells):
        np.testing.assert_array_equal(x.data, y.data)
    assert _regions(a) == _regions(b)
    assert sorted(a.point_data) == sorted(b.point_data)
    for name in a.point_data:
        np.testing.assert_array_equal(a.point_data[name], b.point_data[name])
    assert sorted(a.cell_data) == sorted(b.cell_data)
    for name in a.cell_data:
        for x, y in zip(a.cell_data[name], b.cell_data[name]):
            np.testing.assert_array_equal(x, y)


@pytest.mark.parametrize("path", FIXTURES, ids=[p.name for p in FIXTURES])
def test_engines_read_the_same_mesh(path):
    _same(meshioplusplus.febio.read(path), py_febio.read(path))


def test_every_solid_is_valid_in_meshio_order():
    """Mid-edge nodes on midpoints, hex27 face centres on their faces, positive
    volumes: the FEBio node order was undone correctly."""
    mesh = meshioplusplus.read(MESHES / "all_elements_v40.feb")
    solids = [b for b in mesh.cells if b.type.rstrip("0123456789") != "line"]
    solids = [b for b in solids if b.type not in ("quad", "triangle")]
    assert [b.type for b in solids] == [
        "tetra",
        "tetra10",
        "wedge",
        "wedge15",
        "pyramid",
        "pyramid13",
        "hexahedron",
        "hexahedron20",
        "hexahedron27",
    ]
    for block in solids:
        for row in block.data:
            assert _is_valid(mesh.points[row], block.type), block.type


def test_what_each_mesh_section_becomes(engine):
    mesh = engine.read(MESHES / "all_elements_v40.feb")
    got = {(r.name, r.kind): r for r in mesh.regions}
    # Every <Elements> block is a cell region tagged with its domain's material.
    assert got[("part_hex27", "cell")].tag == 9
    assert got[("part_quad4", "cell")].dim == 2
    # Named <Nodes> blocks are node sets, as in FEBio; ranges expand.
    assert len(got[("SolidNodes", "point")].entries) == 108
    assert got[("first_ten", "point")].entries.tolist() == list(range(10))
    # A surface on solid faces is a side region; one off them its own block.
    assert got[("hex_base", "side")].entries.tolist() == [[6, 4]]
    floating = got[("floating", "cell")]
    assert mesh.cells[10].type == "triangle" and floating.entries.tolist() == [10]
    assert [mesh.cells[k].type for k in (11, 12)] == ["line", "line"]
    assert got[("hex_edge", "cell")].entries.tolist() == [11]
    assert got[("springs", "cell")].entries.tolist() == [12]
    # MeshData: NaN outside the set the array is defined on.
    temperature = mesh.point_data["temperature"]
    hex8 = mesh.cells[6].data[0]
    assert temperature[hex8[:4]].tolist() == [10, 20, 30, 40]
    assert np.isnan(np.delete(temperature, hex8[:4])).all()
    assert mesh.cell_data["fiber"][6].tolist() == [[1, 0, 0]]
    assert np.isnan(mesh.cell_data["fiber"][0]).all()


@pytest.mark.parametrize("version", ["v25", "v30"])
def test_older_specs(engine, version):
    mesh = engine.read(MESHES / f"block_{version}.feb")
    got = {r.name: r for r in mesh.regions}
    assert (got["left"].tag, got["right"].tag) == (1, 2)
    assert got["fixed"].entries.tolist() == [0, 1, 2, 3]
    assert got["both"].entries.tolist() == [0, 1]
    assert got["top"].kind == "side" and len(got["top"].entries) == 2
    assert [a.tolist() for a in mesh.cell_data["thickness"]] == [[0.5], [1.0]]


def test_writers_write_the_same_bytes(tmp_path):
    for path in FIXTURES:
        mesh = meshioplusplus.read(path)
        mesh.point_data.clear()
        mesh.cell_data.clear()
        _core.febio_write(str(tmp_path / "cpp.feb"), mesh)
        py_febio.write(tmp_path / "py.feb", mesh)
        assert (tmp_path / "cpp.feb").read_bytes() == (tmp_path / "py.feb").read_bytes()


@pytest.mark.parametrize("path", FIXTURES, ids=[p.name for p in FIXTURES])
def test_round_trip(engine, path, tmp_path):
    mesh = meshioplusplus.read(path)
    mesh.point_data.clear()
    mesh.cell_data.clear()
    engine.write(tmp_path / "out.feb", mesh)
    back = engine.read(tmp_path / "out.feb")
    np.testing.assert_allclose(back.points, mesh.points)
    assert [(c.type, c.data.tolist()) for c in back.cells] == [
        (c.type, c.data.tolist()) for c in mesh.cells
    ]
    # Names and memberships survive; the tags (material ids) are renumbered.
    before = {(r.name, r.kind): np.asarray(r.entries).tolist() for r in mesh.regions}
    after = {(r.name, r.kind): np.asarray(r.entries).tolist() for r in back.regions}
    assert after == before


def test_what_the_writer_makes_of_foreign_blocks(tmp_path):
    """Boundary triangles on a tet become a <Surface> (read back as a side
    region); lines on its edges an <Edge>; a lone two-node line a
    <DiscreteSet>; vertices are dropped."""
    points = np.array([[0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1], [3, 3, 3]], float)
    mesh = meshioplusplus.Mesh(
        points,
        [
            ("tetra", np.array([[0, 1, 2, 3]])),
            ("triangle", np.array([[0, 2, 1]])),
            ("line", np.array([[0, 1]])),
            ("line", np.array([[3, 4]])),
            ("vertex", np.array([[4]])),
        ],
    )
    meshioplusplus.write(tmp_path / "out.feb", mesh)
    text = (tmp_path / "out.feb").read_text()
    assert '<Surface name="Surface2">' in text
    assert '<Edge name="Edge3">' in text
    assert '<DiscreteSet name="DiscreteSet4">' in text
    assert "<delem>4,5</delem>" in text
    assert '<SolidDomain name="Part1" mat="Part1"/>' in text
    back = meshioplusplus.read(tmp_path / "out.feb")
    assert [c.type for c in back.cells] == ["tetra", "line", "line"]
    assert {(r.name, r.kind) for r in back.regions} == {
        ("Part1", "cell"),
        ("Surface2", "side"),
        ("Edge3", "cell"),
        ("DiscreteSet4", "cell"),
    }


def test_errors(engine, tmp_path):
    bad = tmp_path / "bad.feb"
    bad.write_text('<febio_spec version="1.2"><Geometry/></febio_spec>')
    with pytest.raises(meshioplusplus.ReadError, match="1.2"):
        engine.read(bad)
    bad.write_text(
        '<febio_spec version="4.0"><Mesh><Nodes><node id="1">0,0,0</node></Nodes>'
        '<Elements type="tet4" name="p"><elem id="1">1,2,3,4</elem></Elements>'
        "</Mesh></febio_spec>"
    )
    with pytest.raises(meshioplusplus.ReadError, match="undefined node 2"):
        engine.read(bad)
    bad.write_text(
        '<febio_spec version="4.0"><Mesh><Nodes>'
        + "".join(f'<node id="{i}">{i},0,0</node>' for i in range(1, 6))
        + '</Nodes><Elements type="tet5" name="p"><elem id="1">1,2,3,4,5</elem>'
        "</Elements></Mesh></febio_spec>"
    )
    with pytest.raises(meshioplusplus.ReadError, match="tet5"):
        engine.read(bad)
    assert engine.read(bad, lenient=True).cells[0].type == "tetra"
    bad.write_text(
        '<febio_spec version="3.0"><Geometry><Part name="a"/></Geometry></febio_spec>'
    )
    with pytest.raises(meshioplusplus.ReadError, match="Part"):
        engine.read(bad)


def test_mesh_from_another_file(engine, tmp_path):
    (tmp_path / "driver.feb").write_text(
        '<febio_spec version="4.0"><Module type="solid"/>'
        f'<Mesh from="{MESHES / "block_v30.feb"}"/></febio_spec>'
    )
    # A 4.0 driver pulling the <Mesh> of a 3.0 file, as FEBio's `from=` does.
    mesh = engine.read(tmp_path / "driver.feb")
    assert [c.type for c in mesh.cells] == ["hexahedron", "hexahedron"]


def test_sniffed_and_registered():
    assert meshioplusplus.sniff_format(MESHES / "block_v25.feb") == "febio"
    out = meshioplusplus.formats()
    assert out["extensions"][".feb"] == ["febio"]

File: tests/cpp/test_febio.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// Tests for the FEBio `.feb` reader (spec 2.5/3.0/4.0) and spec-4.0 writer.

// External includes
#include <gtest/gtest.h>

// System includes
#include <cmath>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/formats/febio.hpp"
#include "meshioplusplus/operations/sniff.hpp"
#include "meshioplusplus/read_options.hpp"
#include "meshioplusplus/region.hpp"

using namespace meshioplusplus;
using namespace mt;

namespace {

std::string put(const std::string& rText) {
    const std::string path = temp_path(".feb");
    std::ofstream(path, std::ios::binary) << rText;
    return path;
}

std::string slurp(const std::string& rPath) {
    std::ifstream in(rPath, std::ios::binary);
    std::stringstream ss;
    ss << in.rdbuf();
    return ss.str();
}

const meshioplusplus::Region* find_region(const Mesh& rMesh, const std::string& rName,
                                          RegionKind Kind) {
    for (std::size_t r = 0; r < rMesh.NumRegions(); ++r)
        if (rMesh.Region(r).mName == rName && rMesh.Region(r).mKind == Kind)
            return &rMesh.Region(r);
    return nullptr;
}

// A unit hex27 in FEBio's numbering (FECore FEHex27): corners, bottom, top and
// vertical mid-edges, then face centres y-, x+, y+, x-, z-, z+ and the centre.
const double kHex27[27][3] = {
    {0, 0, 0},   {1, 0, 0},   {1, 1, 0},   {0, 1, 0},   {0, 0, 1},   {1, 0, 1},    {1, 1, 1},
    {0, 1, 1},   {.5, 0, 0},  {1, .5, 0},  {.5, 1, 0},  {0, .5, 0},  {.5, 0, 1},   {1, .5, 1},
    {.5, 1, 1},  {0, .5, 1},  {0, 0, .5},  {1, 0, .5},  {1, 1, .5},  {0, 1, .5},   {.5, 0, .5},
    {1, .5, .5}, {.5, 1, .5}, {0, .5, .5}, {.5, .5, 0}, {.5, .5, 1}, {.5, .5, .5},
};

std::string hex27_spec40() {
    std::ostringstream s;
    s << "<?xml version=\"1.0\"?>\n<febio_spec version=\"4.0\">\n<Module type=\"solid\"/>\n"
      << "<Material><material id=\"7\" name=\"steel\" type=\"isotropic elastic\"/></Material>\n"
      << "<Mesh>\n<Nodes name=\"all\">\n";
    for (int i = 0; i < 27; ++i)  // sparse ids: 10, 20, ...
        s << "<node id=\"" << 10 * (i + 1) << "\">" << kHex27[i][0] << ',' << kHex27[i][1] << ','
          << kHex27[i][2] << "</node>\n";
    s << "</Nodes>\n<Elements type=\"hex27\" name=\"brick\">\n<elem id=\"5\">";
    for (int i = 0; i < 27; ++i)
        s << (i ? "," : "") << 10 * (i + 1);
    s << "</elem>\n</Elements>\n"
      << "<NodeSet name=\"bottom\">10:40:10</NodeSet>\n"
      << "<ElementSet name=\"all_elems\">5</ElementSet>\n"
      << "<Surface name=\"base\"><quad4 id=\"1\">10,40,30,20</quad4></Surface>\n"
      << "<Surface name=\"loose\"><tri3 id=\"1\">10,20,70</tri3></Surface>\n"
      << "</Mesh>\n<MeshDomains><SolidDomain name=\"brick\" mat=\"steel\"/></MeshDomains>\n"
      << "<MeshData><NodeData name=\"t\" node_set=\"bottom\" data_type=\"scalar\">"
      << "<node lid=\"2\">5</node></NodeData></MeshData>\n</febio_spec>\n";
    return s.str();
}

}  // namespace

TEST(Febio, ReadsSpec40WithHex27InFebioOrder) {
    const std::string path = put(hex27_spec40());
    EXPECT_EQ(sniff_format(path), "febio");
    const Mesh mesh = read_febio(path);
    ASSERT_EQ(mesh.NumCellBlocks(), 2u);  // the brick, then the loose surface
    EXPECT_EQ(mesh.Cells(0).Type(), "hexahedron27");
    EXPECT_EQ(mesh.Cells(1).Type(), "triangle");
    // meshio++ (VTK) hexahedron27 face centres: x-, x+, y-, y+, z-, z+.
    const double centres[6][3] = {{0, .5, .5}, {1, .5, .5}, {.5, 0, .5},
                                  {.5, 1, .5}, {.5, .5, 0}, {.5, .5, 1}};
    const NDArray& conn = mesh.Cells(0).Conn();
    for (int f = 0; f < 6; ++f)
        for (int d = 0; d < 3; ++d)
            EXPECT_DOUBLE_EQ(detail::read_double(
                                 mesh.Points(),
                                 static_cast<std::size_t>(detail::read_int(conn, 20 + f) * 3 + d)),
                             centres[f][d])
                << "face centre " << f;
    const auto* brick = find_region(mesh, "brick", RegionKind::Cell);
    ASSERT_NE(brick, nullptr);
    EXPECT_EQ(brick->mTag, 7);  // the domain's material id
    EXPECT_EQ(find_region(mesh, "bottom", RegionKind::Point)->NumEntries(), 4u);
    const auto* base = find_region(mesh, "base", RegionKind::Side);
    ASSERT_NE(base, nullptr);
    EXPECT_EQ(base->Entries()[1], 4);  // hexahedron face 4 = z-
    EXPECT_NE(find_region(mesh, "loose", RegionKind::Cell), nullptr);
    const NDArray& t = mesh.PointData("t");
    EXPECT_EQ(detail::read_double(t, 1), 5.0);  // lid 2 of `bottom` = node id 20
    EXPECT_TRUE(std::isnan(detail::read_double(t, 0)));
}

TEST(Febio, WritesAndReadsBack) {
    const Mesh mesh = read_febio(put(hex27_spec40()));
    const std::string out = temp_path(".feb");
    write_febio(out, mesh);
    const std::string text = slurp(out);
    EXPECT_NE(text.find("<febio_spec version=\"4.0\">"), std::string::npos);
    // `all_elems` and `brick` both cover the block exactly: the first in name
    // order names it, the other is written as an <ElementSet>.
    EXPECT_NE(text.find("<Elements type=\"hex27\" name=\"all_elems\">"), std::string::npos);
    EXPECT_NE(text.find("<SolidDomain name=\"all_elems\" mat=\"all_elems\"/>"),
              std::string::npos);
    EXPECT_NE(text.find("<ElementSet name=\"brick\">1</ElementSet>"), std::string::npos);
    // The loose triangle is not on a solid face: a shell, not a surface.
    EXPECT_NE(text.find("<ShellDomain name=\"loose\" mat=\"loose\"/>"), std::string::npos);
    const Mesh back = read_febio(out);
    ASSERT_EQ(back.NumCellBlocks(), 2u);
    EXPECT_EQ(cell_rows(back), cell_rows(mesh));
    expect_points_close(mesh, back, 0.0);
    for (const char* name : {"brick", "loose", "all_elems"})
        EXPECT_NE(find_region(back, name, RegionKind::Cell), nullptr) << name;
    EXPECT_NE(find_region(back, "base", RegionKind::Side), nullptr);
    EXPECT_NE(find_region(back, "bottom", RegionKind::Point), nullptr);
}

TEST(Febio, ReadsSpec25Geometry) {
    const std::string path =
        put("<febio_spec version=\"2.5\"><Module type=\"solid\"/><Geometry>"
            "<Nodes><node id=\"1\">0,0,0</node><node id=\"2\">1,0,0</node>"
            "<node id=\"3\">0,1,0</node><node id=\"4\">0,0,1</node></Nodes>"
            "<Elements type=\"tet4\" mat=\"3\" name=\"t\"><elem id=\"1\">1,2,3,4</elem></Elements>"
            "<NodeSet name=\"n\"><node id=\"1\"/><node id=\"4\"/></NodeSet>"
            "<ElementSet name=\"e\"><elem id=\"1\"/></ElementSet>"
            "</Geometry><MeshData><ElementData var=\"h\" elem_set=\"e\"><elem lid=\"1\">2</elem>"
            "</ElementData></MeshData></febio_spec>");
    const Mesh mesh = read_febio(path);
    EXPECT_EQ(find_region(mesh, "t", RegionKind::Cell)->mTag, 3);
    EXPECT_EQ(find_region(mesh, "n", RegionKind::Point)->NumEntries(), 2u);
    EXPECT_EQ(detail::read_double(mesh.CellData("h", 0), 0), 2.0);
}

TEST(Febio, Errors) {
    EXPECT_THROW(read_febio(put("<febio_spec version=\"1.2\"/>")), ReadError);
    EXPECT_THROW(read_febio(put("<notfebio/>")), ReadError);
    const std::string tet5 =
        put("<febio_spec version=\"4.0\"><Mesh><Nodes>"
            "<node id=\"1\">0,0,0</node><node id=\"2\">1,0,0</node><node id=\"3\">0,1,0</node>"
            "<node id=\"4\">0,0,1</node><node id=\"5\">.2,.2,.2</node></Nodes>"
            "<Elements type=\"tet5\" name=\"p\"><elem id=\"1\">1,2,3,4,5</elem></Elements>"
            "</Mesh></febio_spec>");
    EXPECT_THROW(read_febio(tet5), ReadError);
    ReadOptions lenient;
    lenient.mLenient = true;
    EXPECT_EQ(read_febio(tet5, lenient).Cells(0).Type(), "tetra");
    EXPECT_THROW(read_febio(put("<febio_spec version=\"4.0\"><Mesh><Elements type=\"tet4\" "
                                "name=\"p\"><elem id=\"1\">1,2,3,4</elem></Elements></Mesh>"
                                "</febio_spec>")),
                 ReadError);
    Mesh polygon;
    polygon.AssignPoints(points_from({{0, 0, 0}, {1, 0, 0}, {1, 1, 0}, {0, 1, 0}, {0, 2, 0}}));
    polygon.AddCellBlock("polygon5", conn_from({{0, 1, 2, 3, 4}}));
    EXPECT_THROW(write_febio(temp_path(".feb"), polygon), WriteError);
}
